include/deal.II/grid/connectivity.h

```diff
@@ -0,0 +1,1477 @@
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_tria_connectivity_h
+#define dealii_tria_connectivity_h
+
+#include <deal.II/base/config.h>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+
+namespace internal
+{
+  namespace TriangulationImplementation
+  {
+    struct CellTypeBase
+    {
+      virtual ~CellTypeBase() = default;
+
+      virtual unsigned int
+      n_entities(const unsigned int d) const
+      {
+        Assert(false, ExcNotImplemented());
+        (void)d;
+
+        return 0;
+      }
+
+      virtual dealii::ArrayView<const unsigned int>
+      vertices_of_entity(const unsigned int d, const unsigned int e) const
+      {
+        Assert(false, ExcNotImplemented());
+        (void)d;
+        (void)e;
+
+        return dealii::ArrayView<const unsigned int>();
+      }
+
+      virtual unsigned int
+      type_of_entity(const unsigned int d, const unsigned int e) const
+      {
+        Assert(false, ExcNotImplemented());
+        (void)d;
+        (void)e;
+
+        return -1;
+      }
+
+      virtual unsigned int
+      n_lines_of_surface(const unsigned int line) const
+      {
+        Assert(false, ExcNotImplemented());
+        (void)line;
+
+        return 0;
+      }
+
+      virtual unsigned int
+      nth_line_of_surface(const unsigned int line,
+                          const unsigned int face) const
+      {
+        Assert(false, ExcNotImplemented());
+        (void)line;
+        (void)face;
+
+        return 0;
+      }
+
+      virtual const std::array<unsigned int, 2> &
+      vertices_of_nth_line_of_surface(const unsigned int line,
+                                      const unsigned int face) const
+      {
+        Assert(false, ExcNotImplemented());
+        (void)line;
+        (void)face;
+
+        const static std::array<unsigned int, 2> table = {};
+
+        return table;
+      }
+    };
+
+
+
+    struct CellTypeLine : public CellTypeBase
+    {
+      dealii::ArrayView<const unsigned int>
+      vertices_of_entity(const unsigned int d,
+                         const unsigned int e) const override
+      {
+        (void)e;
+
+        if (d == 1)
+          {
+            static const std::array<unsigned int, 2> table = {{0, 1}};
+
+            AssertDimension(e, 0);
+
+            return dealii::ArrayView<const unsigned int>(table);
+          }
+
+        Assert(false, ExcNotImplemented());
+
+        return dealii::ArrayView<const unsigned int>();
+      }
+
+      virtual unsigned int
+      type_of_entity(const unsigned int d, const unsigned int e) const override
+      {
+        (void)e;
+
+        if (d == 1)
+          return 1;
```

Comment:
Can we use an `enum` for the entity type rather than a plain integer? In the long run I guess it is very cumbersome to remember that `1` refers to a line or `6` to a pyramid (or was it a wedge?).